eng/common/pipelines/templates/steps/publish-1es-artifact.yml

```diff
@@ -0,0 +1,30 @@
+# This step is used to prevent duplication of artifact publishes when there is an issue that would prevent the overall success of the job.
+# Ensuring that we only publish when successful (and two a differently named artifact otherwise) will allow easy retry on a build pipeline
+# without running into the "cannot override artifact" failure when we finally do get a passing run.
+
+# ArtifactName - The name of the artifact in the "successful" case.
+# ArtifactPath - The path we will be publishing.
+# CustomCondition - Used if there is additional logic necessary to prevent attempt of publish.
+
+parameters:
+  ArtifactName: ''
+  ArtifactPath: ''
+  CustomCondition: true
+
+steps:
+  - pwsh: |
+      Write-Host "##vso[task.setvariable variable=PublishArtifactName;]${{ parameters.ArtifactName }}"
+    condition: and(succeeded(), ${{ parameters.CustomCondition }})
+    displayName: Set Artifact Name
+
+  - pwsh: |
+      Write-Host "##vso[task.setvariable variable=PublishArtifactName;]${{ parameters.ArtifactName }}-FailedAttempt$(System.JobAttempt)"
+    condition: and(failed(), ${{ parameters.CustomCondition }})
+    displayName: Set Failed Artifact Name
+
+  - task: 1ES.PublishPipelineArtifact@1
+    condition: and(succeeded(), ${{ parameters.CustomCondition }})
```

Comment:
Shouldn't we remove or at least update this condition to be failedOrSucceeded()?

Reply:
Ah, yes, good catch. I'll do another PR. FYI @scbedd but shouldn't be necessary for your testing.

Reply:
https://github.com/Azure/azure-sdk-tools/pull/7810

Reply:
I would suggest succededOrFailed here as well in your pwsh step to set the artifact name. https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#succeededorfailed